goosebit/models.py

```diff
@@ -38,3 +45,65 @@ class Rollout(Model):
     paused = fields.BooleanField(default=False)
     success_count = fields.IntField(default=0)
     failure_count = fields.IntField(default=0)
+
+
+class FirmwareCompatibility(Model):
+    id = fields.IntField(primary_key=True)
+    hw_model = fields.CharField(max_length=255)
+    hw_revision = fields.CharField(max_length=255)
+
+
+class FirmwareUpdate(Model):
+    id = fields.IntField(primary_key=True)
+    uri = fields.CharField(max_length=255)
+    size = fields.BigIntField()
+    hash = fields.CharField(max_length=255)
+    version = fields.CharField(max_length=255)
+    compatibility = fields.ManyToManyField(
+        "models.FirmwareCompatibility",
+        related_name="updates",
+        through="update_compatibility",
+    )
+
+    @classmethod
+    async def latest(cls):
+        updates = await cls.all()
+        return sorted(
+            updates,
+            key=lambda x: semver.Version.parse(x.version),
+            reverse=True,
+        )[0]
+
+    @property
+    def path(self):
+        return Path(url2pathname(unquote(urlparse(self.uri).path)))
+
+    @property
+    def local(self):
+        return urlparse(self.uri).scheme == "file"
+
+    def generate_chunk(self, request: Request) -> list:
```

Comment:
Starting to overload the model classes? Looks like an API related aspect.

Reply:
Moved to take `FirmwareUpdate` as a parameter.